cpp/src/arrow/compute/kernels/aggregate_basic_internal.h

```diff
@@ -71,41 +72,51 @@ struct SumImpl : public ScalarAggregator {
 
   void MergeFrom(KernelContext*, KernelState&& src) override {
     const auto& other = checked_cast<const ThisType&>(src);
+    this->length += other.length;
     this->count += other.count;
     this->sum += other.sum;
   }
 
   void Finalize(KernelContext*, Datum* out) override {
-    if (this->count == 0) {
+    if (this->count < options.min_count) {
       out->value = std::make_shared<OutputType>();
     } else {
       out->value = MakeScalar(this->sum);
     }
   }
 
+  size_t length = 0;
   size_t count = 0;
   typename SumType::c_type sum = 0;
+  ScalarAggregateOptions options;
 };
 
 template <typename ArrowType, SimdLevel::type SimdLevel>
 struct MeanImpl : public SumImpl<ArrowType, SimdLevel> {
   void Finalize(KernelContext*, Datum* out) override {
-    if (this->count == 0) {
+    if (this->count == 0 || this->count < options.min_count) {
```

Comment:
Should min_counts check replace `this->count == 0 `, instead of adding it?

Reply:
Would that distinguish between `this->count == 0` and `this->count == 1`?
What check would you propose here? I might be overthinking something here :)

Reply:
> Would that distinguish between `this->count == 0` and `this->count == 1`?

That would depend on the value of `min_count` ? (for `min_count=0`, it would not distinguish both, for `min_count=1`, it would)

Reply:
How about the case `this->count == options.min_count == 0`?

Reply:
That would then take the `else` clause, and depending on the skip_nulls option, that might result in NaN from dividing 0/0 ?

Reply:
Ah got it. Yeah that makes sense now.